locale.c: Don't do asymmetric back out on failure

khwilliamson opened the pull request:
This fixes #23159

When something goes wrong doing locale-aware string collation, the code attempts to carry on as well as can be expected.  prior to this commit the backout code was asymmetric, trying to undo things that had not been done.  This happened when the failure was early on.
     
In the case of this ticket, the platform has a defective locale that was detectable before getting very far along.
     
The solution adopted here is to jump to a different label for those early failures that does less backout than for later failures.

* This set of changes does not require a perldelta entry.

Leont commented:
It's not clear to me if `free_me`, `sans_nuls` and `sans_highs` will be correctly freed when an error occurs.

tonycoz commented:
There's a few paths where they can leak.

sevan commented:
```
All tests successful.
Elapsed: 2660 sec
u=22.73  s=11.66  cu=2088.93  cs=219.98  scripts=2748  tests=1180795
```

khwilliamson committed:
mem_collxfrm(): Return early if locale collation not sane

This changes a subsidiary function's return value from void to bool,
returning false if it finds the locale doesn't have sane collation.

The calling function is changed to check this, and give up immediately
if the locale isn't sane.

khwilliamson committed:
mem_collxfrm(): White space, comments, only

The previous commit removed a block; so can outdent

khwilliamson commented:
I've reexamined this commit, and don't see anything wrong with it.  It does look to me that the first two goto's can be replaced by returns, as nothing has been allocated yet.   But I don't see any leak paths. @tonycoz can you provide a case

mauke commented on locale.c:
```diff
@@ -10090,7 +10090,7 @@ Perl_mem_collxfrm_(pTHX_ const char *input_string,
         DEBUG_L(PerlIO_printf(Perl_debug_log,
                       "mem_collxfrm_: Couldn't malloc %zu bytes\n", xAlloc));
-        goto bad;
+        goto bad_no_strxfrm;
```
This looks like a potential memory leak to me. At this point, `sans_highs`, `sans_nuls`, and `free_me` may have been allocated, but `bad_no_strxfrm` skips over the `free()` calls.

khwilliamson replied:
Somehow I got it into my head that that label had code to do the necessary cleanup, and was blind to the fact that it didn't.  But with the latest push, it should now.

khwilliamson committed:
locale.c: Change name of macro

This is in preparation for the next commit where it will be split out
to be a stand-alone macro.

khwilliamson committed:
locale.c: Don't do asymmetric back out on failure

This fixes #23519

When something goes wrong doing locale-aware string collation, the code
attempts to carry on as well as can be expected.  Prior to this commit
the backout code was asymmetric, trying to undo things that had not been
done.  This happened when the failure was early on.

In the case of this ticket, the platform has a defective locale that was
detectable before getting very far along.

The solution adopted here is to jump to a different label for those
early failures that does less backout than for later failures.

sevan commented:
Test suite still passes on Tiger/i386 with the latest version of the patch applied.
```
All tests successful.                                                                                                                      
Elapsed: 2294 sec                                                                                                                          
u=23.85  s=10.62  cu=1848.69  cs=199.30  scripts=2658  tests=1314307 
```
